services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,453 @@
+const assert = require('assert');
+const slugid = require('slugid');
+const _ = require('lodash');
+const taskcluster = require('taskcluster-client');
+const forge = require('node-forge');
+const crypto = require('crypto');
+const path = require('path');
+const fs = require('fs');
+
+const auth = require('@azure/ms-rest-nodeauth');
+const ComputeManagementClient = require('@azure/arm-compute').ComputeManagementClient;
+const NetworkManagementClient = require('@azure/arm-network').NetworkManagementClient;
+
+const {ApiError, Provider} = require('./provider');
+const {CloudAPI} = require('./cloudapi');
+
+class AzureProvider extends Provider {
+
+  constructor({
+    providerConfig,
+    fakeCloudApis,
+    ...conf
+  }) {
+    super(conf);
+    this.configSchema = 'config-azure';
+    this.providerConfig = providerConfig;
+    this.fakeCloudApis = fakeCloudApis;
+  }
+
+  async setup() {
+    let {
+      clientId,
+      secret,
+      domain,
+      subscriptionId,
+      apiRateLimits = {},
+      _backoffDelay = 1000,
+    } = this.providerConfig;
+
+    // Azure SDK has builtin retry logic: https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific
+    // compute rate limiting: https://docs.microsoft.com/en-us/azure/virtual-machines/troubleshooting/troubleshooting-throttling-errors
+    const cloud = new CloudAPI({
+      types: ['get', 'query', 'list'],
+      apiRateLimits,
+      intervalDefault: 100 * 1000, // Intervals are enforced every 100 seconds
+      intervalCapDefault: 2000, // The calls we make are all limited 20/sec so 20 * 100 are allowed
+      monitor: this.monitor,
+      providerId: this.providerId,
+      errorHandler: ({err, tries}) => {
+        if (err.code === 429) { // too many requests
+          return {backoff: _backoffDelay * 50, reason: 'rateLimit', level: 'notice'};
+        } else if (err.code >= 500) { // For 500s, let's take a shorter backoff
+          return {backoff: _backoffDelay * Math.pow(2, tries), reason: 'errors', level: 'warning'};
+        }
+        // If we don't want to do anything special here, just throw and let the
+        // calling code figure out what to do
+        throw err;
+      },
+    });
+    this._enqueue = cloud.enqueue.bind(cloud);
+
+    // load microsoft intermediate certs from disk
+    // TODO: we should download the intermediate certs,
```

Comment:
add reference to bug here

Reply:
Maybe a note here as well as to why there are 1/2/4/5 and no 3? Otherwise it looks like a typo